src/json.hpp

```diff
@@ -1398,119 +1398,68 @@ constexpr T static_const<T>::value;
-        {
-            buffer_pos += 3;
-            processed_chars += 3;
-        }
-    }
-
-    ~cached_input_stream_adapter() override
+    ~input_stream_adapter() override
     {
-        // clear stream flags
-        is.clear();
-        // We initially read a lot of characters into the buffer, and we may
-        // not have processed all of them. Therefore, we need to "rewind" the
-        // stream after the last processed char.
-        is.seekg(start_position);
-        is.ignore(static_cast<std::streamsize>(processed_chars));
-        // clear stream flags
+        // clear stream flags; we use underlying streambuf I/O, do not maintain ifstream flags
         is.clear();
     }
-
-    int get_character() override
+    explicit input_stream_adapter(std::istream& i)
+        : is(i)
     {
-        // check if refilling is necessary and possible
-        if (buffer_pos == fill_size and not eof)
+        // Ignore Byte Order Mark at start of input
+        int c;
+        if (( c = get_character() ) == 0xEF )
         {
-            fill_buffer();
-
-            // check and remember that filling did not yield new input
-            if (fill_size == 0)
+            if (( c = get_character() ) == 0xBB )
             {
-                eof = true;
-                return std::char_traits<char>::eof();
+                if (( c = get_character() ) == 0xBF )
+                {
+                    return; // Ignore BOM
+                }
+                else if ( c != std::char_traits<char>::eof() )
+                {
+                    is.unget();
+                }
+                is.putback( '\xBB' );
             }
-
-            // the buffer is ready
-            buffer_pos = 0;
+            else if ( c != std::char_traits<char>::eof() )
+            {
+                is.unget();
+            }
+            is.putback( '\xEF' );
+        }
+        else if ( c != std::char_traits<char>::eof() )
+        {
+            is.unget(); // Not BOM.  Process as usual.
         }
-
-        ++processed_chars;
-        assert(buffer_pos < buffer.size());
-        return buffer[buffer_pos++] & 0xFF;
     }
 
-    std::string read(std::size_t offset, std::size_t length) override
+    int get_character() override
     {
-        // create buffer
-        std::string result(length, '\0');
-
-        // save stream position
-        const auto current_pos = is.tellg();
-        // save stream flags
-        const auto flags = is.rdstate();
-
-        // clear stream flags
-        is.clear();
-        // set stream position
-        is.seekg(static_cast<std::streamoff>(offset));
-        // read bytes
-        is.read(&result[0], static_cast<std::streamsize>(length));
-
-        // reset stream position
-        is.seekg(current_pos);
-        // reset stream flags
-        is.setstate(flags);
-
-        return result;
+        int c = is.rdbuf()->sbumpc(); // Avoided for performance: int c = is.get();
```

Comment:
Is there a specific reason to not return `char ` instead of `int`? That's what I assumed from the function name.

Reply:
std::streambuf I/O deals in ints, returning either traits_type::eof() (-1, usually) or the buffer item, converted to an int.  I'm not certain why; perhaps to not assume that the underlying data can only be 8-bit?

Reply:
Indeed, I was thinking about our interface: do we want to expose int or char for callers of `get_character`?

I believe we should `static _cast` between int/char while dealing with the `std::streambuf`, and only exposing `char`